### ISSUE TITLE
Include detailed comments on few of the new React Hook examples

### ISSUE THREAD

**Issue by desmap** [drcmda/react-spring#371]
It's nice to see that react-spring already implemented React Hooks and further provides tons of examples! This is great and I highly appreciate your work. 

But I still can't get a good understanding of how react-spring actually works. I am trying to implement a gallery with React Hooks where five pictures cross-fade infinitely. I took one of the new examples (the simple-transition one), adjusted it a bit and got something working but not really well. And tbh, I don't have a clue what is going on.

The code: https://codesandbox.io/s/0yv00j8j40

Some questions (and they show that I have not the slightest grasp):

1. What's the difference between Spring and Transition? Can I also use Springs for my use case?
2. The Transition doc says 'it takes a list of items of any type, and their keys', does 'list' implies an array? What is meant by 'keys'? The indexes of array items? What if I have just one item and no array?
3. However, what are 'keys', for what does your Transition component needs them and why should my app care? What is meant by 'The latter defaults to item => item' in the Transition documentation?
4. Could you elaborate on 'The full signature looks like (item, state, index) => props => ReactNode' with some examples? Where do you pass those args? In the examples I saw that you pass 'item', state as individual objects ('from'. 'enter', 'leave') and no index as key but as value from items; so it's not like in the documentation
5. Does react-spring works together with CSS-in-JS generated CSS from modules like emotion?

Maybe it's already a huge step for us if you could just include extensive comments in one or two examples and maybe write a post on the fundamentals of react-spring as a bonus. It looks like an awesome lib and I need to fully understand its foundation in order to use it properly and not just copy and pasting snippets.

**Comment by drcmda** [comment]
Here's your demo cleaned up: https://codesandbox.io/embed/morr206pv8

**Comment by desmap** [comment]
@drcmda, many thanks for the quick response and cleaning up my code! This was super helpful: I didn't know the trick that function args keep their value in a closure and hence, I don't need useRef anymore. Also the item array looks much nicer now (was along day yesterday).

One thing I noticed in Chrome Inspector is that the fading-out div are still in the DOM and kind of polluting it. Since I was not sure if this is intended or not, I forked your version and tested it with react-pose where the leaving divs are also removed from the DOM, look: https://codesandbox.io/s/2v783085j0

This was also a good opportunity to compare briefly both modules: Our slideshow example ran on react-spring constantly with 30fps while react-pose was jumping between 10 - 40 fps but there was no difference for the eye (both had the same windows size and where running independently). react-spring does allow to set the spring's tension and friction while react-pose does allow the stiffness only. However, you can also choose a tween instead of a spring with react-pose. Latter also allows to set different configs for enter and exit transition. Regarding the API, I had to create my own map function with react-pose.

**Comment by desmap** [comment]
Just wondering again: How does 'set' does know about the state in a closure? I just set the initial state to 2 instead of 0 and then, set knew about the '2' despite we haven't passed the initial state to set anywhere.

**Comment by drcmda** [comment]
> One thing I noticed in Chrome Inspector is that the fading-out div are still in the DOM and kind of polluting it. Since I was not sure if this is intended or not

No, that's a bug in the useTransition code, @ruggedy could you take a look? 

> However, you can also choose a tween instead of a spring with react-po

You can tween, though that's not our focus. But just do config: { duration: 1000, easing: ... } if you need it

>  Latter also allows to set different configs for enter and exit transitions.

You can do this. config can receive a function that gets all info, the item, enter/leave/update, etc. Here's a very complex example doing this: https://codesandbox.io/embed/7mqy09jyq

**Comment by desmap** [comment]
@drcmda, nice, you released a new version, that was quick 👍

**Comment by ooloth** [comment]
> Latter also allows to set different configs for enter and exit transitions.

> You can do this very easily. config can receive a function that gets all contextual info: the item, enter/leave/update, etc. Here's a very complex example doing this: https://codesandbox.io/embed/7mqy09jyq (`config: (item, state) => (state === 'leave' ? ...`).

Good to know!
